### PR TITLE
Consolidate model types from TCGC

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/4784
Fixes https://github.com/Azure/autorest.csharp/issues/4304

# Description

Align with the model definition on TCGC.

Known issues:
1. The derived types with discriminator will still have the discriminator property defined on them. For example:
```
@discriminator("kind")
model Base {
	kind: string;
}

model Derived extends Base {
	kind: "a";
}
```
we will have the property `kind` on type `Derived` with a different type of literal. It is not causing anything changed in the generated code, but leading to a lot of extra enums because we are wrapping literals into enums. Made a workaround on the PR to fix this.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by ArcturusZhang** [comment]
`Models-TypeSpec` has some issue, need to investigate

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 namespace AutoRest.CSharp.Common.Input
 {
-    internal record InputModelType(string Name, string? Namespace, string? Accessibility, string? Deprecated, string? Description, InputModelTypeUsage Usage, IReadOnlyList<InputModelProperty> Properties, InputModelType? BaseModel, IReadOnlyList<InputModelType> DerivedModels, string? DiscriminatorValue, string? DiscriminatorPropertyName, InputDictionaryType? InheritedDictionaryType, IReadOnlyList<InputType>? ArgumentTypes = null)
+    internal record InputModelType(string Name, string? Namespace, string? Access, string? Deprecation, string? Description, InputModelTypeUsage Usage, IReadOnlyList<InputModelProperty> Properties, InputModelType? BaseModel, IReadOnlyList<InputModelType> DerivedModels, string? DiscriminatorValue, InputModelProperty? DiscriminatorProperty, IReadOnlyDictionary<string, InputModelType> DiscriminatedSubtypes, InputType? AdditionalProperties, IReadOnlyList<InputType>? ArgumentTypes = null)
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:10]: it turns out that we have to keep the `DerivedModels` property here.
We need this to know whether this model has children, if so, we need to add `protected` to raw data field.

Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:10]: In that case, we still need to keep the field order in `InputModelType` that the `BaseModel` should keep `BaseModel` and `Properties` as the last two item in `InputModelType`

**Review by chunyu3 — APPROVED** [review]

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs (hunk 1 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:10]: as you could see, everything works fine.
The order is "property deserialization order", not the order you are seeing here.

**Pull request merged**